compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs

```diff
@@ -748,8 +748,27 @@ convertBind env (name, x)
     -- lifting where the lifted version of `f` happens to be `name`.)
     -- This workaround should be removed once we either have a proper lambda
     -- lifter or DAML-LF supports local recursion.
-    | (as, Let (Rec [(f, Lam v y)]) (Var f')) <- collectBinders x, f == f'
-    = convertBind env $ (,) name $ mkLams as $ Lam v $ Let (NonRec f $ mkVarApps (Var name) as) y
+    --
+    -- NOTE(SF): Due to issue #7953, this has been modified to allow for
+    -- additional (nonrecursive) let bindings between the top-level
+    -- arguments and the letrec. In particular,
+    --
+    --  > name = \@a_1 ... @a_n ->
+    --  >    let { x_1 = e_1 ; ... ; x_m = e_m } in
+    --  >    letrec f = \v -> y in f
+    --
+    -- is rewritten to
+    --
+    --  > name = \@a_1 ... @a_n ->
+    --  >    let { x_1 = e_1 ; ... ; x_m = e_m } in
+    --  >    \v -> let f = name @a_1 ... @a_n in y
+    --
+    | let (args, body1) = collectBinders x
+    , let (lets, body2) = collectNonRecLets body1
+    , Let (Rec [(f, Lam v y)]) (Var f') <- body2
```

Comment:
What happens if we have a multi-param lambda?

Reply:
Currying should handle it fine. I guess it's not optimal performance-wise to break a lambda apart like this, but I didn't want to change the existing transformation in this PR.

Reply:
That works as well. The transformation would insert the `let f = name ... in` after the first parameter. We want to insert it as soon as possible in the parameter list to avoid any shadowing/name capture issues but we also need to have at least one lambda before it such that we don't get a recursive top-level value. Does make sense?

@sofiafaro-da We should probably also check that `f` and `v` are distinct, otherwise the binders for them don't commute and rewriting is illegal.

Reply:
Wondering off the top of my head: How would one trigger a shadowing here? Wouldn't shadowing prevent a recursive binding in the first place? E.g.

```
f f = f 10
```

is not recursive, even if it compiles.

Reply:
Makes sense, thanks!

Reply:
@sofiafaro-da You're totally right. 😅